pkg/util/kubernetes/autoscalers/datadogexternal.go

```diff
@@ -101,51 +103,56 @@ func (p *Processor) queryDatadogExternal(ddQueries []string, timeWindow time.Dur
 		if existingPoint, found := processedMetrics[ddQueries[queryIndex]]; found {
 			if existingPoint.Valid {
 				existingPoint.Valid = false
-				existingPoint.Timestamp = currentTime.Unix()
+				existingPoint.Timestamp = currentTimeUnix
 				existingPoint.Error = errors.New("multiple series found. Please change your query to return a single serie")
 				processedMetrics[ddQueries[queryIndex]] = existingPoint
 			}
 			continue
 		}
 
-		// Use on the penultimate bucket, since the very last bucket can be subject to variations due to late points.
-		var skippedLastPoint bool
-		var point Point
-		// Find the most recent value.
+		// As we batch queries to Datadog API, all returned series return the same number of points, aligned to the smallest rollup value.
+		// This means that a lot of points can be `nil`.
+		// What we want is to find the two most recent points that are not `nil` and use the penultimate one if present, otherwise the last one.
+		var matchedPoint *datadog.DataPoint
 		for i := len(serie.Points) - 1; i >= 0; i-- {
 			if serie.Points[i][value] == nil {
-				// We need this as if multiple metrics are queried, their points' timestamps align this can result in empty values.
 				continue
 			}
-			// We need at least 2 points per window queried on batched metrics.
-			// If a single sparse metric is processed and only has 1 point in the window, use the value.
-			if !skippedLastPoint && len(serie.Points) > 1 {
-				// Skip last point unless the query window only contains one valid point
-				skippedLastPoint = true
-				continue
-			}
-			point.Value = *serie.Points[i][value]                       // store the original value
-			point.Timestamp = int64(*serie.Points[i][timestamp] / 1000) // Datadog's API returns timestamps in s
-			point.Valid = true
 
-			m := fmt.Sprintf("%s{%s}", *serie.Metric, *serie.Scope)
-			processedMetrics[ddQueries[queryIndex]] = point
+			if matchedPoint == nil {
+				matchedPoint = &serie.Points[i]
+			} else if matchedPoint != nil {
+				// Penuultimate point found, we can stop here.
+				matchedPoint = &serie.Points[i]
+				break
+			}
```

Comment:
```suggestion
			matchedPoint = &serie.Points[i]
			if matchedPoint != nil {
				// Penuultimate point found, we can stop here.
				matchedPoint = &serie.Points[i]
				break
			}
```
Could it be simplified by that ?
Also, why do we want the penultimate instead of the last one ?

Reply:
It could but I think it makes the intent a bit less clear. We use the penultimate point because the last point may not be complete (leading to scaling on incorrect data)